[FEATURE] VM 인스턴스 연동 구현

moonhyeonmin opened the pull request:
### PR 본문
- Resolved: #130

## 📋 개요
AgenticCP에 AWS EC2 인스턴스 관리 기능을 헥사고날 아키텍처로 구현하고, 실제 AWS 연동·검증까지 완료했습니다. 입력 검증/에러 매핑/로깅을 정비해 운영 안정성을 높였습니다.

## 🏗️ 헥사고날 아키텍처 구현
### 1) 포트(Ports)
- Ec2ManagementPort: 인스턴스 생성/시작/중지/재부팅/종료/삭제
- Ec2DiscoveryPort: 인스턴스 단건/목록 조회
- CredentialProviderPort: 자격증명 해결(환경변수/CLI/역할 기반 확장 포인트)

### 2) 어댑터(Adapters)
- AwsEc2ManagementAdapter: AWS EC2 Run/Start/Stop/Reboot/Terminate/Delete
- AwsEc2Mapper: AWS SDK 모델 ↔ 도메인 모델 매핑
- AuditEventAdapter / TracingAdapter / CredentialProviderAdapter 연동

### 3) 유스케이스(Use Case)
- Ec2UseCaseService: 비즈니스 로직 및 감사 로깅
- Ec2PortRouter: 프로바이더별 포트 라우팅(AWS)

### 4) 컨트롤러(Interface)
- Ec2Controller: REST 엔드포인트 제공(목록/상세/생성/수정/생명주기/삭제)

## 🔄 주요 흐름
- 생성: Controller → UseCase → PortRouter → ManagementAdapter → AWS EC2
- 조회: Controller → UseCase → PortRouter → AWS EC2

## ⚙️ Configuration
### AwsEc2Config.java
- `DefaultCredentialsProvider` 기반으로 AWS CLI/환경변수/IAM 역할 자동 탐색
- 리전/엔드포인트 오버라이드 지원(LocalStack 등)
- 실패 시 명확한 로그와 안전한 생성 처리

```java
@Bean
public Ec2Client ec2Client() {
  return Ec2Client.builder()
      .region(Region.of(region))
      .credentialsProvider(DefaultCredentialsProvider.create())
      .build();
}
```

## ✅ 요청/검증/매핑 개선
- Ec2CreateRequest: `@Jacksonized` + `@Min(1)`로 `minCount/maxCount` 기본값·검증 확정
- Ec2Controller: `@Valid` 적용으로 입력 검증 강화
- AwsEc2Mapper: 보정 로직 제거, 순수 매핑만 수행
- 목록 조회 파라미터: `@RequestParam`로 명시적 바인딩(page/size/instanceId/instanceName/state/instanceType/availabilityZone)

## 🧭 엔드포인트 요약
- GET `/api/api/v1/ec2/instances` (필터·페이징)
- GET `/api/api/v1/ec2/instances/{instanceId}`
- POST `/api/api/v1/ec2/instances` (imageId, instanceType 등)
- POST `/api/api/v1/ec2/instances/{instanceId}/start|stop|reboot|terminate`
- DELETE `/api/api/v1/ec2/instances/{instanceId}` (본문 없이 삭제 가능)

## 🔐 보안/운영
- JWT 인증 요구(엔드포인트 공개하지 않음)
- 감사 로깅(AuditLogger), MDC/분산 트레이싱 연계
- CloudErrorTranslator: 공급자 오류를 일관된 비즈니스 예외로 변환
- Mock 코드 전면 제거, 실제 AWS 연동만 유지

## 🧪 테스트
- 수동 검증: Swagger/cURL 통해 실제 AWS 호출 검증
<img width="1452" height="1622" alt="image" src="https://github.com/user-attachments/assets/e2262ec3-6773-46a6-97f8-a2297d371ac5" />
- 시나리오: 목록/상세/생성(ubuntu 22.04 AMI)/시작/중지/재부팅/종료/삭제
- 문제 해결: InvalidAMIID, min/max 기본값 미반영, 삭제 시 path 변수 누락 등 케이스 처리

## 💡 운영 가이드
- 자격증명: `aws configure` 후 실행(또는 환경변수), `AWS_ENABLED=true`
- 리전: `ap-northeast-2` 기본(변경 가능)
- 로그: `logs/agenticcp-core.log`에서 AWS SDK 원인 메시지 확인

## Breaking Changes
- 없음(EC2 도메인 내 요청 DTO 검증 추가만 적용)

## 체크리스트
- [x] 헥사고날 구조(Ports/Adapters/UseCase/Controller)
- [x] 실제 AWS 연동 및 Swagger 검증
- [x] 입력 검증/에러 처리/로깅 정비
- [x] Mock 제거 및 설정 단순화

요약
- EC2 전 기능(조회/생성/수정/생명주기/삭제)을 헥사고날로 구현하고, 실제 AWS 환경에서 동작을 검증했습니다. 입력 검증과 로깅/오류 매핑을 정비해 운영 안정성을 확보했습니다.

moonhyeonmin committed:
feat: AWS SDK v2 의존성 및 EC2 기본 설정 추가

- pom.xml에 AWS SDK v2 의존성 추가 (ec2, aws-core, auth)
- @Value 어노테이션을 사용한 AwsConfig 구현
- application.yml과 application-local.yml에 AWS 설정 추가
- 환경변수 기반 설정 지원
- LocalStack 테스트용 엔드포인트 오버라이드 포함
- 조건부 로딩으로 EC2 클라이언트 빈 설정 (aws.enabled=true)

설정 내용:
- 리전 설정 (기본값: us-east-1)
- 자격증명 프로바이더 (정적 또는 기본)
- LocalStack 테스트용 엔드포인트 오버라이드
- 연결 및 읽기 타임아웃 설정

참고: AWS SDK 로드까지 EC2 클라이언트 생성 임시 비활성화

moonhyeonmin committed:
feat: Define EC2 domain models (Query, CreateRequest, UpdateRequest, DeleteRequest)

- Ec2Query.java - EC2 조회 조건 모델
- Ec2CreateRequest.java - EC2 생성 요청 모델
- Ec2UpdateRequest.java - EC2 수정 요청 모델
- Ec2DeleteRequest.java - EC2 삭제 요청 모델

moonhyeonmin committed:
feat: EC2 관리 포트 인터페이스 정의

- Ec2ManagementPort 인터페이스 구현
- EC2 인스턴스 전체 생명주기 관리 기능 정의
- 인스턴스 조회, 생성, 수정, 삭제, 생명주기 관리 메서드 포함
- 태그 관리 및 상태 확인 기능 추가
- 헥사고날 아키텍처의 핵심 포트(계약) 구현
- 상세한 JavaDoc으로 계약 명시 및 예외 상황 정의

주요 기능:
- 인스턴스 조회 (listInstances, getInstance)
- 생명주기 관리 (create, start, stop, reboot, terminate, delete)
- 인스턴스 수정 (updateInstance)
- 태그 관리 (addTags, removeTags, getTags)
- 상태 확인 (getInstanceStatus, waitForInstanceStatus)

moonhyeonmin committed:
feat: AWS EC2 매퍼 실제 SDK 통합 구현 완료

- 임시 Map 기반 구현을 실제 AWS SDK 클래스로 교체
- AWS Instance와 CloudResource 간 완전한 데이터 변환 구현
- 모든 EC2 작업 매핑 추가 (조회, 생성, 수정, 삭제)
- 인스턴스 타입별 CPU/메모리 매핑 및 메타데이터 추출 지원
- 데이터 변환을 위한 적절한 에러 처리 및 로깅 추가
- AwsConfig를 실제 Ec2Client.Builder 패턴 사용하도록 업데이트
- 매퍼 기능에 대한 단위 테스트 추가

변경사항:
- AwsEc2Mapper: Instance, DescribeInstancesRequest, RunInstancesRequest 등 완전한 AWS SDK 통합
- AwsConfig: 적절한 빌더 패턴으로 실제 Ec2Client 생성 활성화
- AwsEc2MapperSimpleTest: 기본 매퍼 기능 테스트

EC2 관리를 위한 핵사고날 아키텍처의 데이터 변환 계층을 완성합니다.

moonhyeonmin committed:
feat: AWS EC2 어댑터 구현 완료 - 모든 EC2 관리 기능 제공

- Ec2ManagementPort 인터페이스의 모든 메서드 구현
- AWS SDK와 도메인 계층을 연결하는 핵사고날 아키텍처 어댑터 완성
- 인스턴스 조회, 생성, 수정, 삭제, 생명주기 관리 기능 제공
- 태그 관리 및 상태 확인 기능 구현
- CloudErrorTranslator를 통한 에러 처리
- ProviderScoped 인터페이스 구현으로 AWS 제공자 식별

구현된 기능:
- listInstances: EC2 인스턴스 목록 조회 (페이징 지원)
- getInstance: 단일 인스턴스 조회
- createInstance: 새 인스턴스 생성
- startInstance/stopInstance/rebootInstance: 생명주기 관리
- terminateInstance/deleteInstance: 인스턴스 종료/삭제
- updateInstance: 인스턴스 정보 수정
- addTags/removeTags/getTags: 태그 관리
- getInstanceStatus/waitForInstanceStatus: 상태 확인 및 대기

핵사고날 아키텍처의 어댑터 계층이 완성되어 AWS EC2와의 통합이 가능합니다.

moonhyeonmin committed:
feat: EC2 포트 라우터 구현 완료 - 제공업체별 어댑터 동적 선택

- Ec2PortRouter 클래스 구현으로 제공업체별 EC2 어댑터 관리
- ProviderScoped 인터페이스를 통한 어댑터 자동 등록 및 식별
- 런타임에 제공업체 타입에 따라 적절한 어댑터 선택
- 지원되지 않는 제공업체에 대한 명확한 에러 처리
- 제공업체 지원 여부 확인 및 등록된 포트 정보 조회 기능

구현된 기능:
- ec2(ProviderType): 제공업체별 EC2 포트 반환
- getSupportedProviders(): 지원되는 제공업체 목록 조회
- isProviderSupported(): 제공업체 지원 여부 확인
- getPortCount(): 등록된 포트 개수 조회

핵사고날 아키텍처의 라우터 계층이 완성되어 다중 클라우드 환경에서
동적으로 적절한 어댑터를 선택할 수 있습니다.

moonhyeonmin committed:
feat: EC2 인스턴스 발견을 위한 유스케이스 서비스 구현

- 모든 CRUD 작업을 포함한 Ec2UseCaseService 완전 구현
- 모든 EC2 작업에 대한 포괄적인 감사 로깅 추가
- 인스턴스 생명주기 관리 구현 (시작, 중지, 재부팅, 종료)
- 태그 관리 기능 추가 (태그 추가, 제거, 조회)
- 상태 확인 및 대기 기능 포함
- Mock 검증을 포함한 포괄적인 단위 테스트 추가
- 적절한 관심사 분리와 함께 핵사고날 아키텍처 원칙 준수
- 적절한 트랜잭션 관리 및 로깅과 함께 서비스 패턴 적용

moonhyeonmin committed:
test: EC2 인스턴스 생성 기능에 대한 포괄적인 테스트 추가

- Ec2UseCaseServiceCreateTest 클래스 추가
- AwsEc2MapperCreateTest 클래스 추가
- 다양한 생성 시나리오 테스트 (기본, 전체옵션, 태그포함, 예외처리)
- AWS SDK RunInstancesRequest 변환 로직 검증
- Mock 검증을 통한 포트 호출 확인
- 감사 로그 기록 검증

moonhyeonmin committed:
test: EC2 인스턴스 생성 기능에 대한 포괄적인 테스트 추가

- Ec2UseCaseServiceCreateTest 클래스 추가
- AwsEc2MapperCreateTest 클래스 추가
- 다양한 생성 시나리오 테스트 (기본, 전체옵션, 태그포함, 예외처리)
- AWS SDK RunInstancesRequest 변환 로직 검증
- Mock 검증을 통한 포트 호출 확인
- 감사 로그 기록 검증

moonhyeonmin committed:
feat: EC2 REST API 컨트롤러 구현

- EC2 인스턴스 관리를 위한 포괄적인 REST API 엔드포인트 추가
- 모든 CRUD 작업 지원: 조회, 생성, 수정, 삭제
- 생명주기 관리 구현: 시작, 중지, 재부팅, 종료
- 태그 관리 추가: 태그 추가, 제거, 조회
- 상태 모니터링 포함: 상태 확인, 상태 대기
- 모든 엔드포인트에 대한 완전한 Swagger/OpenAPI 문서화
- 적절한 HTTP 메서드를 사용한 RESTful 설계 원칙 준수
- 포괄적인 로깅 및 에러 처리 추가
- 비즈니스 로직을 위한 Ec2UseCaseService와 통합
- 인스턴스 목록 조회를 위한 페이징 지원
- 성공 및 에러 응답을 적절히 처리

API 엔드포인트:
- GET    /api/v1/ec2/instances                    # 인스턴스 목록 조회
- GET    /api/v1/ec2/instances/{id}               # 인스턴스 조회
- POST   /api/v1/ec2/instances                    # 인스턴스 생성
- PUT    /api/v1/ec2/instances/{id}               # 인스턴스 수정
- DELETE /api/v1/ec2/instances/{id}               # 인스턴스 삭제
- POST   /api/v1/ec2/instances/{id}/start         # 인스턴스 시작
- POST   /api/v1/ec2/instances/{id}/stop          # 인스턴스 중지
- POST   /api/v1/ec2/instances/{id}/reboot        # 인스턴스 재부팅
- POST   /api/v1/ec2/instances/{id}/terminate     # 인스턴스 종료
- POST   /api/v1/ec2/instances/{id}/tags          # 태그 추가
- DELETE /api/v1/ec2/instances/{id}/tags          # 태그 제거
- GET    /api/v1/ec2/instances/{id}/tags          # 태그 조회
- GET    /api/v1/ec2/instances/{id}/status        # 상태 확인
- POST   /api/v1/ec2/instances/{id}/wait          # 상태 대기

이로써 EC2 관리를 위한 핵사고날 아키텍처 구현이 완료되어,
외부 클라이언트를 위한 완전한 REST API 인터페이스를 제공합니다.

moonhyeonmin committed:
test: EC2 계약 테스트 및 통합 테스트 추가

- 포트 인터페이스 계약 검증을 위한 Ec2ManagementContractTest 추가
- 어댑터 기본 기능 테스트를 위한 AwsEc2ManagementAdapterTest 추가
- REST API 엔드포인트 테스트를 위한 Ec2ControllerTest 추가
- EC2 관리 계층의 모든 테스트 커버리지 완성
- 포괄적인 Mock 테스트로 포트 인터페이스 계약 검증
- 어댑터 의존성 주입 및 제공업체 타입 식별 테스트
- 컨트롤러 HTTP 메서드 및 요청/응답 처리 테스트
- 적절한 에러 처리 및 엣지 케이스 커버리지 보장

테스트 커버리지:
- 포트 인터페이스 계약 테스트 (Ec2ManagementContractTest)
- 어댑터 통합 테스트 (AwsEc2ManagementAdapterTest)
- 컨트롤러 API 테스트 (Ec2ControllerTest)
- 기존 유스케이스 서비스 테스트 (Ec2UseCaseServiceTest*)
- 기존 매퍼 테스트 (AwsEc2MapperTest*)
- 기존 라우터 테스트 (Ec2PortRouterTest)

이로써 EC2 관리 시스템의 포괄적인 테스트 스위트가 완성되어,
모든 아키텍처 계층에서의 신뢰성과 유지보수성이 확보되었습니다.

moonhyeonmin committed:
fix: 테스트 환경에서 Ec2Client Bean 누락으로 인한 ApplicationContext 로딩 실패 해결

- MockAdaptersConfig에 Mock Ec2Client와 AwsEc2Mapper Bean 추가
- 테스트 환경에서 AWS SDK 의존성 없이도 정상 동작하도록 개선
- AgenticCpCoreApplicationTests.contextLoads 테스트 정상 통과 확인

Resolves: ApplicationContext 로딩 실패 문제

moonhyeonmin committed:
feat: 실제 구현 환경에 맞게 클라우드 어댑터들 구현

- AuditEventAdapter: 기존 감사 시스템과 연동하여 일관된 감사 로깅 제공
  * AuditLogger와 AuditEventDto 활용
  * AuditResourceType.CLOUD_PROVIDER 사용
  * 결과에 따른 자동 심각도 결정 및 속성 추출

- CredentialProviderAdapter: 암호화된 자격증명 관리 및 캐시 기능 구현
  * EncryptionService를 통한 암호화된 자격증명 처리
  * application.yml 설정 기반 AWS 자격증명 관리
  * 메모리 기반 캐시로 성능 최적화
  * 테넌트별 자격증명 및 캐시 무효화 기능

- TracingAdapter: MDC 기반 분산 추적 시스템 구현
  * ThreadLocal 기반 스팬 컨텍스트 관리
  * 계층적 부모-자식 스팬 관계 지원
  * MDC 연동으로 로그에 추적 정보 자동 포함
  * 설정 기반 추적 기능 활성화/비활성화

모든 어댑터가 프로젝트의 실제 구조와 패턴에 맞게 구현되어
향후 외부 시스템과의 연동 준비 완료

moonhyeonmin committed:
feat(ec2): 실제 AWS 연동 완료 및 EC2 API 안정화
AwsEc2Config: DefaultCredentialsProvider 적용(AWS CLI 자격증명 사용), 설정 단순화
application-local.yml/start-dev.sh: 환경변수 정리(AWS 키 제거), CLI 방식 권장
Ec2CreateRequest: @Jacksonized + @Min(1)로 minCount/maxCount 기본값·검증 보장
Ec2Controller: @Valid 적용으로 입력 검증 강화
AwsEc2Mapper: min/max 하드코딩 보정 제거, 순수 매핑 유지
에러 로깅 개선: AWS 원인 메시지 노출(InvalidAMIID.NotFound 등)로 디버깅 용이
EC2 삭제 엔드포인트 사용 가이드 반영(경로 변수 필수)
Refs: feature/#130

moonhyeonmin committed:
feat: CapabilityGuard 적용 및 EC2 capability 등록, 테스트 호환성 개선
Ec2CapabilityConfig: AWS|EC2|INSTANCE capability 등록(start/stop/terminate/tagging 지원)
Ec2UseCaseService:
CapabilityGuard 선검증 적용(start/stop/reboot/terminate/delete/tags)
reboot는 START/STOP 모두 지원 검증
테스트 환경 호환을 위해 CapabilityGuard 선택 주입(@Autowired(required=false)) 처리
전체 동작 수동 검증(엔드포인트 정상, 가드 동작 확인)

moonhyeonmin committed:
feat(capability, ec2): 생성/수정에 CapabilityGuard 적용 및 플래그 추가

- CspCapability에 supportsCreate, supportsUpdate 플래그 추가
- CapabilityGuard.Operation에 CREATE/UPDATE 추가 및 검증 로직 반영
- Ec2CapabilityConfig에서 EC2 INSTANCE에 create/update 지원 등록
- Ec2UseCaseService의 create/update에 ensureSupported 적용
- 린트/컴파일 이상 없음

Refs: #130

hyobin-yang commented on src/main/java/com/agenticcp/core/domain/cloud/controller/VmController.java:
```diff
@@ -0,0 +1,511 @@
+            throw e;
+        }
+    }
+}
```
aws(ec2)에 특화된 컨트롤러가 아닌 다른 CSP로 이 컨트롤러를 통해 vm 등을 생성할 수 있도록 공통화하면 좋을 것 같아요!!그리고 파라미터로 CSP 타입을 받아 서비스 로직으로 보내고 각 CSP 타입에 따라 로직을 처리할 수 있도록 해야 헥사고날 원칙에 위배되지 않을 것 같슴다

hyobin-yang replied:
CSP를 다루는 게 우리 플랫폼에서 가장 중요한 기능 중 하나인 만큼 감사로깅 적용하면 어떨까유

moonhyeonmin replied:
1. 넵 공통화 후, 파라미터로 CSP 타입 받는 방식으로 변경하겠습니다!!
2. 감사로깅 적용해보겠습니다!

hyobin-yang commented on src/main/java/com/agenticcp/core/controller/VmController.java:
```diff
@@ -0,0 +1,511 @@
+        @ApiResponse(responseCode = "400", description = "잘못된 요청"),
+        @ApiResponse(responseCode = "500", description = "서버 오류")
+    })
+    public ResponseEntity<Boolean> waitForInstanceStatus(
```
컨트롤러에서의 반환값이 너무 raw한 것 같아서 정의되어있는 ApiResponse로 감싸서 반환하면 더 좋을 듯합니다!

moonhyeonmin replied:
넵 반영해서 올리겠습니당

hyobin-yang commented on src/main/java/com/agenticcp/core/domain/cloud/capability/CspCapability.java:
```diff
@@ -6,6 +6,8 @@
 public class CspCapability {
+    boolean supportsCreate;
+    boolean supportsUpdate;
     boolean supportsStart;
```
요 두 필드를 추가한 이유를 물어봐도 될까용
왜냐면 저도 이 부분 첨에 고민하다가 CspCapability가 자원의 '구동'과 관련되어 만들어진 클래스가 아닐까 했거든여

moonhyeonmin replied:
> 요 두 필드를 추가한 이유를 물어봐도 될까용 왜냐면 저도 이 부분 첨에 고민하다가 CspCapability가 자원의 '구동'과 관련되어 만들어진 클래스가 아닐까 했거든여

지금 다시 보니까 안 맞는 필드인 것 같네요... 처음에 Capability가 자원의 생성부터 종료까지의 상태를 다 관리하는 것으로 착각해서 추가한 것 같네요!! 수정하겠습니당 🥹

hyobin-yang commented on src/main/java/com/agenticcp/core/domain/cloud/port/model/VmCreateRequest.java:
```diff
@@ -0,0 +1,117 @@
+            ))
+            .build();
+    }
+}
```
이 dto들도 여러 CSP가 사용할 수 있도록 해야할 것 같구 아까 논의한 대로 이 dto들과 동일한 경로에 Command 객체들 생성해서 의존성 끊기 + 각 CSP 매핑하기 하면 구조가 딱딱 잘 떨어질 것 같은ㅎㅎ

hyobin-yang commented on src/main/java/com/agenticcp/core/domain/cloud/service/aws/VmUseCaseService.java:
```diff
@@ -0,0 +1,455 @@
+        try {
+            if (capabilityGuard != null) {
+                capabilityGuard.ensureSupported(ProviderType.AWS, SERVICE_KEY, RESOURCE_TYPE, CapabilityGuard.Operation.UPDATE);
+            }
```
RequiredArgsConstructor 있으니까 capabilityGuard는 null 검사 안 해도 되지 않을까용

hyobin-yang reviewed (commented):
로드가 안 될 정도로 코드가 많네요🥹 수고 많으셨습니다!!👍👍

moonhyeonmin committed:
feat(cloud-vm): EC2 명칭을 VM으로 통일 및 어댑터/설정/테스트 정리
도메인 모델 리네이밍: Ec2* → Vm* (Query/Create/Update/Delete)
애플리케이션 계층: Ec2UseCaseService/Controller → VmUseCaseService/Controller, 라우터 Ec2PortRouter → VmPortRouter
포트/어댑터: Ec2ManagementPort → VmManagementPort, AwsEc2* → AwsVm* (패키지 aws.vm로 이동)
설정: AwsEc2Config → AwsVmConfig
Capability: Ec2CapabilityConfig → VmCapabilityConfig (SERVICE_KEY ‘VM’)
API 경로: /api/v1/ec2 → /api/v1/vms
Resilience4j 키: cloud.aws.ec2.list → cloud.aws.vm.list
테스트 리팩터: 경로/클래스/문자열 VM 기준으로 수정

moonhyeonmin committed:
fix: 반환값 수정 및 경로 수정

moonhyeonmin committed:
feat: CspCapability create, update 필드 제거

moonhyeonmin committed:
refactor: reorganize vm service layer

moonhyeonmin committed:
refactor: VM 포트 구조와 커맨드 분리

moonhyeonmin committed:
refactor: AWS VM 어댑터 역할 분리

moonhyeonmin committed:
chore: capability flag cleanup

moonhyeonmin commented:
@hyobin-yang 피드백 사항 반영 완료했습니다!!! 서영이가 구현해준 부분 머지되면 자격증명도 그때 적용해서 다시 언급드릴게용~~

moonhyeonmin committed:
Merge branch 'merge/hexagonal/aws' of https://github.com/CCCloudPlatform/AgenticCP-Core into feature/#130

moonhyeonmin committed:
feat(VmCommand): VM Command 객체에 세션 자격증명 필드 추가

- VmCreateCommand, VmUpdateCommand, VmDeleteCommand에 CloudSessionCredential session 필드 추가
- 멀티 테넌트 환경에서 자격증명 격리를 위한 구조 개선

moonhyeonmin committed:
feat(cloud): AwsClientConfig 리팩토링

- validateAndCastSession(): 세션 검증 및 AWS 세션 타입 캐스팅
- toSdkCredentials(): 도메인 세션 → AWS SDK 세션 변환
- resolveRegion(): 리전 기본값 처리 (파라미터 > 세션 > us-east-1)
- createEc2Client(CloudSessionCredential, String): 세션 기반 EC2 클라이언트 생성

moonhyeonmin committed:
feat(vm): AwsVmLifecycleAdapter 세션 기반 클라이언트 생성 적용

Management 작업(create, update, delete)에 세션 기반 EC2 클라이언트 생성 패턴 적용

- AwsClientConfig 주입하여 세션 기반 클라이언트 생성
- createInstance: command.getSession()으로 클라이언트 생성
- updateInstance: command.getSession()으로 클라이언트 생성
- deleteInstance: command.getSession()으로 클라이언트 생성
- try-finally 패턴으로 클라이언트 리소스 해제

moonhyeonmin committed:
feat(vm): AwsVmDiscoveryAdapter Execute Around 패턴 적용

Discovery 작업에 JIT(Just-In-Time) 세션 획득 패턴 적용

- 싱글톤 Ec2Client 제거
- AwsClientConfig, AccountCredentialManagementPort 주입
- executeWithEc2Client(): 세션 획득 → 클라이언트 생성 → 작업 실행 → 해제
- listInstances, getInstance, getInstanceStatus, waitForInstanceStatus에 패턴 적용

moonhyeonmin committed:
feat(vm): VmLifecyclePort 시그니처 변경 - 세션 파라미터 추가

모든 생명주기 작업에 세션 자격증명 명시적 전달

Port 인터페이스 변경:
- startInstance(String, CloudSessionCredential)
- stopInstance(String, CloudSessionCredential)
- rebootInstance(String, CloudSessionCredential)
- terminateInstance(String, CloudSessionCredential)

Adapter 변경:
- 싱글톤 Ec2Client 완전 제거
- 모든 메서드가 세션 기반 클라이언트 생성/해제

Service 변경:
- 세션 획득 후 Port 메서드에 명시적 전달

moonhyeonmin committed:
test(vm): 테스트 코드 업데이트 - Phase 6 시그니처 변경 반영

VmLifecyclePort 시그니처 변경 및 Adapter 생성자 변경에 맞춰 테스트 수정

VmUseCaseServiceLifecycleTest:
- verify() 호출에 CloudSessionCredential 파라미터 추가
- doThrow() 설정에 세션 파라미터 매처 추가
- getSession() mock 설정 추가

moonhyeonmin committed:
test: VmManagementContractTest 수정

moonhyeonmin committed:
refactor(vm): ObjectStorageUseCaseService 스타일로 리팩토링

- providerType, accountScope를 메서드 파라미터로 분리
- 감사 로그 및 try-catch 제거
- 로깅 패턴 단순화
- Request 객체에서 providerType/accountScope 제거

moonhyeonmin committed:
refactor(vm): VmCreateRequest/Command를 CSP 중립 필드로 리팩토링

- VmCreateRequest, VmCreateCommand의 AWS 특화 필드를 CSP 중립 필드로 변경
  - imageId → image
  - instanceType → instanceSize
  - keyName → sshKey
  - securityGroupId → networkSecurityId
  - zone 필드 추가

- AwsVmMapper에 CSP 특화 변환 로직 추가
  - 도메인 모델(CSP 중립) → AWS SDK 요청(CSP 특화) 변환
  - zone → placement.availabilityZone 매핑 추가

moonhyeonmin committed:
refactor(vm): API Endpoint 통일

- URL 패턴 변경: /api/v1/vms → /api/v1/cloud/providers/{provider}/accounts/{accountScope}/vms/instances
- @RequestParam → @PathVariable로 providerType, accountScope 전달 방식 변경
- VmCreateRequest, VmDeleteRequest, VmUpdateRequest에 providerType, accountScope 필드 추가
- @Value → @Data 변경 (setter 지원)
- VmUseCaseService: Request 객체에서 providerType, accountScope 추출
- 테스트 코드 업데이트

moonhyeonmin committed:
refactor(vm): 세션 관리 패턴 통일 - Discovery/Tagging에도 getSession() 적용

- VmDiscoveryPort: 모든 메서드에 CloudSessionCredential 세션 파라미터 추가
- VmTaggingPort: 모든 메서드에 CloudSessionCredential 세션 파라미터 추가
- VmUseCaseService: acquireSession() → getSession()으로 변경, 모든 작업에서 세션 전달
- AwsVmDiscoveryAdapter: 세션 파라미터로 클라이언트 생성
- AwsVmTaggingAdapter: 세션 파라미터로 클라이언트 생성
- VmController: Discovery/Tagging 호출에 accountScope 전달

JIT 세션 관리 패턴 준수:
- Service 레벨에서 getSession()으로 세션 획득
- 획득한 세션을 Port/Adapter에 전달
- Redis 캐싱을 통한 세션 재사용

moonhyeonmin commented:
@hyobin-yang 일단, 효빈님 PR과 서영님 Merge 된 자격증명 사안 기반으로 리팩토링 완료했습니다!!

hyobin-yang committed:
Merge branch 'merge/hexagonal/aws' into feature/#130
